### PR TITLE
[BEAM-2670] Deduplicates serializable wrappers over PipelineOptions from Spark, Flink and Apex, and removes buggy SparkRuntimeContext

### PR DESCRIPTION
**Pull request by jkff**
It used a global variable to store the deserialized options, so even if there were several instances of SparkRuntimeContext created with different PipelineOptions, they would all return the same value depending on which one was asked first.

R: @jbonofre or @staslev or @kennknowles or @tgroh or pretty much anyone

### CODE REVIEW

**Comment by jkff** [comment]
Run Spark ValidatesRunner

**Comment by jbonofre** [comment]
Thanks, I'm starting the review.

**Comment by jbonofre** [comment]
Run Spark ValidatesRunner

**Comment by jkff** [comment]
retest this please

**Comment by jkff** [comment]
The Spark ValidatesRunner tests are pre-existing. This commit does fix the JIRA issue though https://issues.apache.org/jira/browse/BEAM-2670.

**Comment by jbonofre** [comment]
retest this please

**Comment by jbonofre** [comment]
Run Spark ValidatesRunner

**Comment by jbonofre** [comment]
Testing locally as well.

**Comment by jkff** [comment]
The Create.ofProvider() test failure looks odd, it doesn't even involve the Spark runner... I'll take a look when I get to the office.

**Comment by jkff** [comment]
Create.ofProvider test is failing because https://github.com/apache/beam/pull/1796 sets a specific version of Jackson. Jackson is used only in SparkRuntimeContext, to serialize and deserialize PipelineOptions. I don't think either of this is correct (we should use just one version of this library, and SparkRuntimeContext shouldn't do its own serialization/deserialization - that's responsibility of something like PipelineOptionsFactory); fixing.

**Comment by jkff** [comment]
Astrologers proclaim the week of the nerd-snipe.

I started by fixing SparkRuntimeContext and then noticed that we have 2 more almost identical classes in Flink and Apex, so I instead created a class in runners-core and deduplicated, and deleted SparkRuntimeContext altogether (along with deleting some other dead code hanging onto it). It also turned out that SparkRuntimeContext didn't call FileSystems.setDefaultPipelineOptions - bug averted, I guess.

**Comment by jkff** [comment]
Run Spark ValidatesRunner

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java [hunk 1]
```diff
@@ -128,6 +129,36 @@ public static PipelineOptions create() {
...
+   * #deserializeFromJson(String)}.
+   *
+   * <p>This method is intended to be used by {@link PipelineRunner runners}. Not intended for use
+   * by pipeline authors. See note "Serialization Of PipelineOptions" on {@link PipelineOptions}. Do
+   * not use this method in your transforms to circumvent the fact that {@link PipelineOptions} is
+   * not serializable or to capture the current value of {@link PipelineOptions} at pipeline
+   * construction time.
+   */
+  public static String serializeToJson(PipelineOptions options) {
+    try {
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java]: I'm not sure how much I like this type signature.

Perhaps this should be put in core construction? I don't think we get all that much from the use of MAPPER (feel free to correct me!)

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SerializablePipelineOptions.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ * FileSystems#setDefaultPipelineOptions(PipelineOptions)} on construction or on deserialization.
+ */
+public class SerializablePipelineOptions implements Serializable {
+  private final String serializedPipelineOptions;
+  private transient PipelineOptions options;
+
+  public SerializablePipelineOptions(PipelineOptions options) {
+    this.serializedPipelineOptions = PipelineOptionsFactory.serializeToJson(options);
+    this.options = options;
+    FileSystems.setDefaultPipelineOptions(options);
```
Comment (tgroh) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SerializablePipelineOptions.java]: This is smelly.

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/SerializablePipelineOptionsTest.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+  @Test
+  public void testIndependence() throws Exception {
+    SerializablePipelineOptions first =
+        new SerializablePipelineOptions(
+            PipelineOptionsFactory.fromArgs("--foo=first").as(MyOptions.class));
+    SerializablePipelineOptions firstCopy = SerializableUtils.clone(first);
+    SerializablePipelineOptions second =
+        new SerializablePipelineOptions(
+            PipelineOptionsFactory.fromArgs("--foo=second").as(MyOptions.class));
```
Comment (tgroh) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/SerializablePipelineOptionsTest.java]: I was expecting a field to be modified on one copy as well, and have a demonstration that the two are independent there as well; maybe include multiple clones?

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/SerializablePipelineOptionsTest.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  @Test
+  public void testSerializationAndDeserialization() throws Exception {
+    PipelineOptions options =
+        PipelineOptionsFactory.fromArgs("--foo=testValue").as(MyOptions.class);
+    SerializablePipelineOptions serializableOptions = new SerializablePipelineOptions(options);
+    SerializablePipelineOptions copy = SerializableUtils.clone(serializableOptions);
+    assertEquals("testValue", serializableOptions.get().as(MyOptions.class).getFoo());
+    assertEquals("testValue", copy.get().as(MyOptions.class).getFoo());
```
Comment (tgroh) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/SerializablePipelineOptionsTest.java]: Perhaps a field that is marked as `@JsonIgnore` does not make it through serialization?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java (hunk 1 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java]: Done, moved into SerializablePipelineOptions (which is in core-construction).

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SerializablePipelineOptions.java (hunk 2 above)
Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SerializablePipelineOptions.java]: Yes, but this is what Apex and Flink runner already did (and Spark probably should have, but didn't); I think it's outside the scope of this PR to fix that smelliness.

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/SerializablePipelineOptionsTest.java (hunk 4 above)
Comment (jkff) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/SerializablePipelineOptionsTest.java]: Done.

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/SerializablePipelineOptionsTest.java (hunk 3 above)
Comment (jkff) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/SerializablePipelineOptionsTest.java]: Done.

**Comment by jkff** [comment]
Thanks, PTAL.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SerializablePipelineOptions.java (hunk 2 above)
Comment (tgroh) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SerializablePipelineOptions.java]: I would like to have a JIRA to remove this, with either subtasks or additional tasks to move the the flink-worker-side code to explicitly do it there; worker processes should own their environment, and should not depend on the time that options are deserialized to set up the defaults.

This is especially the case for the potential of having multiple serialized options - if each sets the 'default' options, we've returned to racy initialization, just elsewhere in the stack.

Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SerializablePipelineOptions.java]: Done, filed https://issues.apache.org/jira/browse/BEAM-2712

**Pull request closed without merging**